server/bleep/src/webserver/answer.rs

```diff
@@ -317,18 +337,18 @@ impl AnswerAPIClient {
             A:",
             snippet.relative_path, snippet.text, self.query
         );
-        let tokens_used = token_count_estimate(&prompt);
-        let max_tokens = 1500_u32.saturating_sub(tokens_used).clamp(100, 2500);
-        self.send(prompt, max_tokens).await
-    }
-}
+        let tokens_used = self.semantic.gpt2_token_count(&prompt);
+        info!(%tokens_used, "input prompt token count");
+        let max_tokens = 4096usize.saturating_sub(tokens_used);
+        if max_tokens == 0 {
+            // our prompt has overshot the token count, log an error for now
+            // TODO: this should propagte to sentry
+            error!(%tokens_used, "prompt overshot token limit");
+        }
 
-// openai claims 4 characters = 1 token for english,
-// but code typically clobbers the token count,
-// we get no more than 3 characters per token.
-//
-// TODO: this is a heuristic, a realistic tokenizer
-// can tune this better, rust_tokenizers::Gpt2Tokenizer?
-fn token_count_estimate(input: &str) -> u32 {
-    input.len().saturating_div(3) as u32
+        // do not let the completion cross 2500 tokens
+        let max_tokens = max_tokens.clamp(1, 500);
```

Comment:
regardless of the amount of wiggle room we have for completions, we set the limit to 500 here. this should hopefully keep the response times down; though in practice, the response times seem to be independent of the completion token count, and dependent on the extent of crowding of the API.